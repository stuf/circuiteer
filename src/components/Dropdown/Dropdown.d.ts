import { FunctionComponent } from 'react';

export interface IChoice<T> {
  value: T;
  label: string;
}

export interface Props<T = string> {
  value: T;
  onChange?: (e: Event) => void;
  choices?: IChoice<T>[];
  disabled?: boolean;
}

export const Dropdown: FunctionComponent<Props>;
