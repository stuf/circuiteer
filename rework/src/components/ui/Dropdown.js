import clsx from 'clsx';

export function Dropdown(props) {
  const { id, className, value, choices = [], onChange, renderChoice } = props;

  const choiceFn = renderChoice || (() => 'foo');

  return (
    <div className={clsx('dropdown', className)}>
      <select {...{ id, onChange, defaultValue: value }}>
        {choices.map((item, i) => choiceFn({ item, i }))}
      </select>
    </div>
  );
}
