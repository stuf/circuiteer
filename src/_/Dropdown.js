import cx from 'classnames';
import * as P from 'prop-types';
import { Listbox, Transition } from '@headlessui/react';

import css from './Dropdown.module.css';

/**
 * @deprecated
 * @param {*} props
 * @returns
 */
function Dropdown(props) {
  const { label, value, choices, onChange } = props;

  return (
    <Listbox {...{ value, onChange }} as="div" className={cx(css.root)}>
      {({ open }) => (
        <>
          <div className="relative">
            <Listbox.Label>{label}</Listbox.Label>

            <Listbox.Button className={css.button}>
              <span className={css.buttonText}>{value.label}</span>
              <span className={css.buttonIconWrap}>
                {/* <SelectorIcon className={css.buttonIcon} aria-hidden="true" /> */}
              </span>
            </Listbox.Button>

            <Transition
              show={open}
              enter={css.enter}
              enterFrom={css.enterFrom}
              enterTo={css.enterTo}
              leave={css.leave}
              leaveFrom={css.leaveFrom}
              leaveTo={css.leaveTo}
            >
              <Listbox.Options static className={css.choices}>
                {choices.map((choice, ix) => (
                  <Listbox.Option
                    key={`choice-${ix}`}
                    value={choice}
                    disabled={choice.unavailable}
                  >
                    {({ active, selected }) => (
                      <>
                        <span
                          className={cx(css.choice, {
                            [css.active]: active,
                            [css.selected]: selected,
                            [css.unavailable]: choice.unavailable,
                          })}
                        >
                          {choice.label}
                        </span>
                        {selected ? (
                          <span className={css.selectedIconWrap}>
                            {/* <CheckIcon
                              aria-hidden="true"
                              className={css.selectedIcon}
                            /> */}
                          </span>
                        ) : null}
                      </>
                    )}
                  </Listbox.Option>
                ))}
              </Listbox.Options>
            </Transition>
          </div>
        </>
      )}
    </Listbox>
  );
}

Dropdown.propTypes = {
  value: P.shape({
    id: P.string,
    label: P.string,
  }),
  choices: P.arrayOf(
    P.shape({
      id: P.string,
      label: P.string,
    }),
  ),
};

Dropdown.defaultProps = {
  value: {},
  choices: [
    { id: 'one', label: 'One' },
    { id: 'two', label: 'Two', unavailable: true },
    { id: 'three', label: 'Three' },
  ],
};

export default Dropdown;
