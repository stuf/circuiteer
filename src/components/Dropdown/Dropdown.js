import * as P from 'prop-types';
import cx from 'classnames';

import { actions } from 'common/util';
import { Icon } from '../Icon';

export function Dropdown(props) {
  const { label, value, onChange, choices, disabled, className } = props;

  return (
    <div
      className={cx(
        className,
        'relative shadow-md rounded-md border-2 bg-white',
        !disabled ? `hover:shadow-lg` : '',
        disabled ? `opacity-50` : '',
      )}
    >
      <span
        className="absolute inset-y-0 right-2 inline-flex items-center"
        aria-hidden="true"
      >
        <Icon name="unfold_more" />
      </span>

      <select
        className={`px-4 py-2 pr-10 appearance-none block w-full bg-transparent`}
        value={value}
        onChange={actions(onChange)}
        disabled={disabled}
      >
        {choices.map((choice, ix) => (
          <option key={ix} value={choice.value}>
            {choice.label}
          </option>
        ))}
      </select>
    </div>
  );
}

Dropdown.propTypes = {
  value: P.any,
  choices: P.arrayOf(
    P.shape({
      value: P.any,
      label: P.string,
    }),
  ).isRequired,
  onChange: P.func,
};

Dropdown.defaultProps = {
  choices: [],
};
