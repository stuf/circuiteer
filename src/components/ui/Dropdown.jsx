import clsx from 'clsx';
import * as P from 'prop-types';

export function Dropdown(props) {
  const { id, className, value, choices = [], onChange, renderChoice } = props;

  const choiceFn = renderChoice || (() => 'foo');

  return (
    <div className={clsx('dropdown', className)}>
      <select {...{ id, onChange, defaultValue: value }}>
        {choices.map((item, i) => choiceFn({ item, i }))}
      </select>
    </div>
  );
}

Dropdown.propTypes = {
  id: P.string.isRequired,
  className: P.string,
  value: P.any,
  choices: P.arrayOf(P.object),
  onChange: P.func,
  renderChoice: P.func.isRequired,
};
