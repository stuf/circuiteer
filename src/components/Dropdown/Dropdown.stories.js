import { action } from '@storybook/addon-actions';

import { Dropdown } from './Dropdown';

export default {
  title: 'UI/Dropdown',
  component: Dropdown,
  argTypes: {},
};

const Template = ({ onChange, ...args }) => {
  return <Dropdown {...args} onChange={action('onChange')} />;
};

export const Basic = Template.bind({});
Basic.args = {
  value: 1,
  choices: [
    { value: 1, label: 'One' },
    { value: 2, label: 'Two' },
    { value: 3, label: 'Three' },
  ],
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};
