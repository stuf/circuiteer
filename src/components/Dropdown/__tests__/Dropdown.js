import { render, fireEvent, logRoles, logDOM } from '@testing-library/react';
import { Dropdown } from '../Dropdown';

test('Dropdown', async () => {
  const onChange = jest.fn();
  const props = {
    choices: [
      { label: 'One', value: 'one' },
      { label: 'Two', value: 'two' },
    ],
    onChange,
  };

  const { container, findByRole } = render(<Dropdown {...props} />);

  const input = await findByRole('combobox');

  fireEvent.change(input, { target: { value: 'two' } });

  expect(onChange).toHaveBeenCalled();
  expect(input.value).toBe('two');

  expect(container).toMatchSnapshot();
});
