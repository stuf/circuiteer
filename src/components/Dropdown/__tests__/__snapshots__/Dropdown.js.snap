// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown works as default 1`] = `
<div>
  <div
    class="relative shadow-md rounded-md border-2 bg-white hover:shadow-lg"
  >
    <span
      aria-hidden="true"
      class="absolute inset-y-0 right-2 inline-flex items-center"
    >
      <span
        class="icon icon--uiflex items-center justify-center"
        style="width: 24px; height: 24px;"
      >
        <span
          class="material-icons material-icons-outlined"
          style="font-size: 24px;"
        >
          unfold_more
        </span>
      </span>
    </span>
    <select
      class="px-4 py-2 pr-10 appearance-none block w-full bg-transparent"
    >
      <option
        value="one"
      >
        One
      </option>
      <option
        value="two"
      >
        Two
      </option>
    </select>
  </div>
</div>
`;
